engines/director/score.cpp

```diff
@@ -1344,123 +1357,273 @@ void Score::playQueuedSound() {
+		_framesStream->seek(_frameOffsets[frameNum]);
+		debugC(7, kDebugLoading, "****** Frame request %d, offset %ld", frameNum, _framesStream->pos());
+
+		if (readOneFrame()) {
+			loadCurrentFrameAction(); // Load actions of current frame!
+			setSpriteCasts(); // Set sprite casts for each sprite in frame!
+			_curFrameNumber = frameNum;
+			return true;
+		} else {
+			warning("Score::loadFrame(): Problem reading previous frame %d, current offset: %ld", frameNum, _framesStream->pos());
+			return false;
+		}
+	}
+
+	// Seek to latest frame
+	_framesStream->seek(_frameOffsets[_frameOffsets.size() - 1]);
+
+	debugC(7, kDebugLoading, "****** Frame request %d, last recorded frame offset %ld", frameNum, _framesStream->pos());
+
+	int currentFrameSkipped = frameNum;
+	// Else read until that specific frame
+	while (_frameOffsets.size() <= (uint)frameNum) {
+		debugC(7, kDebugLoading, "****** Skipping over a frame %d, offset %ld", _frameOffsets.size() - 1, _framesStream->pos());
+
+		if (!readOneFrame(true)) {
+			warning("Score::loadFrame(): Problem reading frame %d, is it outside maximum frames?", frameNum);
+			return false;
 		}
+		currentFrameSkipped++;
+	}
+
+	debugC(7, kDebugLoading, "****** Frame loading %d, offset %ld", frameNum, _framesStream->pos());
+
+	bool isRead = readOneFrame(true);   // Now final read the frame, this is our target frame!
+	if (isRead) {
+		loadCurrentFrameAction();
+		setSpriteCasts();
+		_curFrameNumber = frameNum;
+	} else {
+		warning("Score::loadFrame(): Problem reading frame %d, current offset: %ld", frameNum, _framesStream->pos());
+	}
+
+	return isRead;
+}
+
+// TODO: Not used right now, for future if there are issues with backwards jumping
+// Why? Because director works on concept of "delta" changes between frame, to save
+// space each frame only contains changes from previous frame, so we need to rebuild
+// whole channel data if we were to make a very big jump. This is essential because
+// without it we might have leftover casts/properties which will interfere with channel
+// data after the jump!
+void Score::rebuildChannelData(int frameNum) {
+	// Builds channel data from frame 1 to frame.
+	if (frameNum > (int)_numFrames) {
+		warning("Score::rebuildChannelData(): frameNum %d is greater than total frames %d", frameNum, _numFrames);
+		return;
+	}
 
+	memset(_channelData, 0, kChannelDataSize);
+
+	// Lock variables
+	int curFrameNumber = _curFrameNumber;
+	Frame *frame = _currentFrame;
+	_currentFrame = nullptr; // To avoid later deletion of frame inside renderOneFrame()
+
+	_framesStream->seek(_frameOffsets[1]); // Seek to frame 1
+	for (int i = 1; i < frameNum; i++) {
+		readOneFrame(false);
+	}
+
+	// Unlock variables
+	delete _currentFrame; // Destroy the built frame
+	_curFrameNumber = curFrameNumber;
+	_currentFrame = frame;
+}
+
+bool Score::readOneFrame(bool saveOffset) {
+	uint16 channelSize;
+	uint16 channelOffset;
+
+	if (_framesStream->pos() < _framesStreamSize && !_framesStream->eos()) {
```

Comment:
Instead of wrapping this whole logic into this gigantic `if()`, just inver the condition, e.g.

```C++
if (_framesStream->pos() >= _framesStreamSize || !_framesStream->eos())
  return false;
  
uint16 frameSize = _framesStream->readUint16();
assert(frameSize < _framesStreamSize);
...
```